ext/objspace/objspace_dump.c

```diff
@@ -679,6 +691,61 @@ objspace_dump(VALUE os, VALUE obj, VALUE output)
     return dump_result(&dc);
 }
 
+static void
+shape_i(rb_shape_t *shape, void *data)
+{
+    struct dump_config *dc = (struct dump_config *)data;
+
+    dump_append(dc, "{\"address\":");
+    dump_append_ref(dc, (VALUE)shape);
+
+    dump_append(dc, ", \"type\":\"SHAPE\", \"id\":");
+    dump_append_lu(dc, rb_shape_id(shape));
+
+    dump_append(dc, ", \"parent_id\":");
+    dump_append_lu(dc, shape->parent_id);
+
+    dump_append(dc, ", \"edges\":");
+    dump_append_sizet(dc, rb_shape_edges_count(shape));
+
+    dump_append(dc, ", \"shape_type\":");
+    switch(shape->type) {
+      case SHAPE_ROOT:
+        dump_append(dc, "\"ROOT\"");
+        break;
+      case SHAPE_IVAR:
+        dump_append(dc, "\"IVAR\"");
+        break;
+      case SHAPE_FROZEN:
+        dump_append(dc, "\"FROZEN\"");
+        break;
+      case SHAPE_CAPACITY_CHANGE:
+        dump_append(dc, "\"CAPACITY_CHANGE\"");
+        break;
+      case SHAPE_IVAR_UNDEF:
+        dump_append(dc, "\"IVAR_UNDEF\"");
+        break;
+      case SHAPE_INITIAL_CAPACITY:
+        dump_append(dc, "\"INITIAL_CAPACITY\"");
+        break;
+      case SHAPE_T_OBJECT:
+        dump_append(dc, "\"T_OBJECT\"");
+        break;
+      default:
+        rb_bug("[objspace] unexpected shape type");
+    }
+
+    if (shape->edge_name && ((shape->edge_name & ID_INTERNAL) != ID_INTERNAL)) {
```

Comment:
I think we can probably print the edge name only in the `SHAPE_IVAR` and `SHAPE_IVAR_UNDEF` cases above because that's where it's useful information. Then we won't need `dump_append_id`, and won't need to add another dependency on `ID_INTERNAL`

Reply:
Good idea.

Reply:
So I tried:
```c
      case SHAPE_IVAR_UNDEF:
        dump_append(dc, "\"IVAR_UNDEF\"");

        if ((shape->edge_name & ID_INTERNAL) == ID_INTERNAL) {
            rb_bug("SHAPE_IVAR_UNDEF internal id?");
        }

        dump_append(dc, ",\"edge_name\":");
        dump_append_string_value(dc, rb_sym2str(ID2SYM(shape->edge_name)));
        break;
```

And it does crash in `TestObjSpace#test_load_allocation_path_load`, that is surprising.

Reply:
Ok, I tried to debug this again, but the crash is gone. Not sure what I was doing wrong.

Reply:
Actually, it's back. Here's what the first faulty shape look like:

```
(lldb) p *shape
(rb_shape_t) $3 = {
  edges = 0x0000600000230820
  edge_name = 9231
  next_iv_index = 4
  capacity = 6
  type = '\x01'
  size_pool_index = '\0'
  parent_id = 213
}
```

`type = '\x01'` I assume is `SHAPE_IVAR`.

Reply:
Ok, I tracked it down to:

```ruby
static VALUE
err_init_recv(VALUE exc, VALUE recv)
{
    if (!UNDEF_P(recv)) rb_ivar_set(exc, id_recv, recv);
    return exc;
}
```

So it seems that `ivar` names can perfectly be internal IDs?

Reply:
Here's what it looks like now:

```json
{"address":"0x102a004c0", "type":"SHAPE", "id":38, "parent_id":37, "depth":6, "shape_type":"IVAR","edge_name":"ID_INTERNAL(576)", "edges":1, "memsize":120}
{"address":"0x102a004e0", "type":"SHAPE", "id":39, "parent_id":38, "depth":7, "shape_type":"IVAR","edge_name":"ID_INTERNAL(575)", "edges":1, "memsize":120}
{"address":"0x102a00500", "type":"SHAPE", "id":40, "parent_id":39, "depth":8, "shape_type":"IVAR","edge_name":"args", "edges":1, "memsize":120}
{"address":"0x102a00520", "type":"SHAPE", "id":41, "parent_id":40, "depth":9, "shape_type":"CAPACITY_CHANGE", "capacity":12, "edges":1, "memsize":120}
{"address":"0x102a00540", "type":"SHAPE", "id":42, "parent_id":41, "depth":10, "shape_type":"IVAR","edge_name":"ID_INTERNAL(553)", "edges":1, "memsize":120}
```